### ISSUE TITLE
[Feature Request] Using another GP as mean function

### ISSUE THREAD

**Issue by fonnesbeck** [cornellius-gp/gpytorch#743]
I'm not sure if this is already possible, but I have tried with no success so far: I'm interested in using a GP (or perhaps a DNN or similar model) as a mean function. I have tried subclassing `Mean` and using the predicted mean via:

```
with torch.no_grad(), gpytorch.settings.fast_pred_var():
    return likelihood(model(x)).mean
```

as the output of `forward` but this breaks on fitting the GP:

```
RuntimeError: You must train on the training inputs!
```

I did ensure that the GP and likelihood that is being used for the mean function were in evaluation mode.

Any thoughts here would be welcome.

**Comment by jacobrgardner** [comment]
I know you said you ensured the GP is in eval mode, but I'm a little skeptical because the error you are getting is only ever thrown in training mode:
https://github.com/cornellius-gp/gpytorch/blob/4c71e54f519fe1c51db8d0609c5a88532039f17b/gpytorch/models/exact_gp.py#L221-L229

Is it possible that you set it to eval mode, but when you change the outer GP to train mode it is setting the inner GP to train mode because pytorch sets all submodules to train mode?

Also note that if you want to train through eval mode of a GP, you'll need to do two extra things related to dealing with the caches we compute at test time (e.g., computing and storing K_{xx}^{-1}y for the mean), since these test caches will be invalidated by each step of training:

1. Include a `with gpytorch.settings.detach_test_caches(False)` context around your inner GP call. This will make it possible to backpropagate through the test time caches (e.g., K_{xx}^{-1}y), which you'll need to do for hyperparameter learning but we normally disallow for significant efficiency gains.

2. Clear the caches every time you call `optimizer.step()` or anything else that changes the inner model's hyperparameters. You can do this, e.g., by doing `self._model.train()` and then `self._model.eval()`, because the caches get cleared every time the model gets put in training mode.

**Comment by fonnesbeck** [comment]
OK, thanks I will try your suggestions.

I was putting the model in eval mode in the `__init__` of the `Mean` subclass where I was assigning the model and likelihood I was using to class instance variables.

**Comment by jacobrgardner** [comment]
Yeah you might have to put it in eval mode in the `forward` method of the mean to ensure that even when you swap the outer GP to train mode, when you call the outer GP the inner GP ends up in `eval` mode.

**Comment by fonnesbeck** [comment]
OK, I've got something that *runs* along these lines, but not something that *works*. I've tried to follow the recommendations, and came up with the following:

```python
class GPMean(gpytorch.means.Mean):
    
    def __init__(self, gp):
        super().__init__()
        self.gp = gp
        
    def forward(self, x):
        self.gp.train()
        self.gp.likelihood.train()
        self.gp.eval()
        self.gp.likelihood.eval()
        with torch.no_grad(), gpytorch.settings.fast_pred_var():
            with gpytorch.settings.detach_test_caches(False):
                return self.gp.likelihood(self.gp(x[:, :2])).mean
```

However, this does not produce sensible results. For example, here is the GP that I am using as a prior (fit with a sparse GP): 

![image](https://user-images.githubusercontent.com/81476/61321985-5276d100-a7d2-11e9-9681-c80cc053fefb.png)

but when I use it in the `GPMean` function above as the mean of an exact GP with a small amount of data, it makes a bit of a mess:

![image](https://user-images.githubusercontent.com/81476/61322060-7c2ff800-a7d2-11e9-9d5a-c1bb37df6347.png)

The GP that is used as the prior ends up getting updated while training the outer GP, since running it forward on some test points results in very different predictions before vs after using it in the outer model.

Any obvious things I have done/not done to cause this?

**Comment by jacobrgardner** [comment]
No, that code all looks mostly correct to me actually. The only problems I see are efficiency related: is it your intention to not train the inner GP further (e.g., you have a `torch.no_grad` context)? If so, you won't actually need to toggle between train and eval mode every time -- it's totally fine for the inner GP to cache things if you don't plan on updating its parameters.

Have you verified that the mean you are getting out is sensible in each iteration? As a sanity check, what happens if you manually set the lengthscale very close to zero and make predictions in a `gpytorch.settings.fast_computations(solves=False)` context? Doing these will in some sense "turn off" the outer GP by having K_{XX} -> I and K_{X*X} -> 0, so that the prior mean (e.g., your inner GP) is used everywhere.

**Comment by fonnesbeck** [comment]
No, the inner GP (i.e. the one I am using for the prior mean) should not be updated. Its essentially what the whole population is doing, and I want to use that as a prior so that regions without data in the outer GP have sensible posterior values. Sort of an empirical Bayes approach, if you like.

I will try some of the diagnostics you suggest. Thanks for the help.

**Comment by fonnesbeck** [comment]
Yeah, the GP I am using for the mean function is getting updated by the outer GP, even though I'm calling it in the `no_grad` context. Here are the parameters of the inner GP after 10 iterations:

<img width="825" alt="Screenshot 2019-07-16 15 32 27" src="https://user-images.githubusercontent.com/81476/61327643-05e5c280-a7df-11e9-9236-3ad2f3b2a434.png">

versus prior to training the outer GP:

<img width="796" alt="Screenshot 2019-07-16 15 30 50" src="https://user-images.githubusercontent.com/81476/61327709-2b72cc00-a7df-11e9-8ff7-42f7950b0998.png">

I definitely need these parameters to be static when training the GP that this is the prior mean for (otherwise it isn't really a prior).

**Comment by fonnesbeck** [comment]
OK, so a true "brute force" approach makes it work as intended. Here I actually reload the parameters from the serialized model after every call to `forward()` in the mean function:

```python
class GPMean(gpytorch.means.Mean):
    
    def __init__(self, gp=None):
        super().__init__()
        self.gp = pooled_model = SparseGPRegressionModel(X_all, y_all, inducing_points, pooled_likelihood).cuda()
        self._reset_gp()
        
    def _reset_gp(self):
        self.gp.load_state_dict(torch.load('../models/slider_lhp_rhp_attack.mod'))
        
    def forward(self, x):
        self.gp.eval()
        self.gp.likelihood.eval()
        with torch.no_grad(), gpytorch.settings.fast_pred_var():
            with gpytorch.settings.detach_test_caches(False):
                return self.gp.likelihood(self.gp(x[:, :2])).mean
        self._reset_gp()
```

now you can see that the regions with sparse data reflect the prior mean function:

![image](https://user-images.githubusercontent.com/81476/61328828-80afdd00-a7e1-11e9-9fb2-0df832eaa414.png)

Of course, I'd prefer to just be able to fix the parameters once and keep them unchanged ...
